### PR TITLE
fix: optimize some bit functions

### PR DESCRIPTION
**Pull request by kazuyukitanimura**
## Which issue does this PR close?

Part of #679 and #670

## Rationale for this change

The improvement could be negligible in real use cases, but I see some improvements in micro benchmarks 

## What changes are included in this PR?

Optimizations in some bit functions

## How are these changes tested?

Existing tests

### CODE REVIEW

**Comment by andygrove** [comment]
It would be nice to have a criterion benchmark as part of the PR

**Comment by kazuyukitanimura** [comment]
@andygrove 
```
bit_util/get_value/i32_num_bits_8
                        time:   [781.53 ps 797.98 ps 818.12 ps]
                        change: [-14.395% -11.959% -9.4057%] (p = 0.00 < 0.05)
                        Performance has improved.
bit_util/get_value/i32_num_bits_32
                        time:   [599.30 ps 605.37 ps 613.10 ps]
                        change: [-21.472% -14.798% -7.3131%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 13 outliers among 100 measurements (13.00%)
  13 (13.00%) high mild

bit_util/read_u64       time:   [662.74 ps 667.79 ps 675.09 ps]
                        change: [-44.346% -43.866% -43.140%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) high mild
  2 (2.00%) high severe
bit_util/read_u32       time:   [640.71 ps 641.86 ps 643.37 ps]
                        change: [-84.385% -84.321% -84.263%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 15 outliers among 100 measurements (15.00%)
  4 (4.00%) high mild
  11 (11.00%) high severe
bit_util/get_u32_value  time:   [2.0853 ms 2.0860 ms 2.0868 ms]
                        change: [-28.295% -28.107% -27.945%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild

```
The baseline of `read_u64(...)`/`read_u32(...)` are `read_num_bytes_u64(8, ...)`/`read_num_bytes_u32(4, ...)`
Reverted the changes for `trailing_bits` as no significant differences are observed.

**Review by andygrove — APPROVED** [review]
Those are some nice speedups. Thanks @kazuyukitanimura. I'm not too familiar with this code so it would be good to get another review before merging (perhaps @viirya or @parthchandra?)

File: native/core/src/common/bit.rs
Comment (parthchandra) [inline native/core/src/common/bit.rs:567]: nit: should we use size_of instead of 8?

File: native/core/src/common/bit.rs
Comment (parthchandra) [inline native/core/src/common/bit.rs:570]: If `len > 8` then this will not read `len` bytes will it? Also `read_num_bytes_u64` appears to cover this case in an `unlikely` block (and handles it rather differently).

Comment (kazuyukitanimura) [reply native/core/src/common/bit.rs:570]: If `len > 8` then, this will read 8 bytes only (u64).
`read_num_bytes_u64` has
```
    debug_assert!(size <= src.len());
    if unlikely(src.len() < 8) {
```
So when `size=8` we do not need the `if unlikely()` block

Comment (parthchandra) [reply native/core/src/common/bit.rs:570]: Maybe we can explicitly make the condition `if src.len == 0` 
So 
```
   if src.len == 0 {
     read_u64(buf.as_slice())
   } else {
     read_num_bytes_u64(len, buf.as_slice())
   }
```

Comment (kazuyukitanimura) [reply native/core/src/common/bit.rs:570]: When `if src.len == 0` is true, we cannot read anything if I understand correctly?

File: native/core/src/common/bit.rs
Comment (kazuyukitanimura) [reply native/core/src/common/bit.rs]: updated

**Comment by kazuyukitanimura** [comment]
@parthchandra do you have any more feedbacks?
cc @comphead @huaxingao @viirya

File: native/core/benches/bit_util.rs
Comment (comphead) [inline native/core/benches/bit_util.rs:189]: I would add some comments on magic numbers

File: native/core/src/common/bit.rs
Comment (comphead) [inline native/core/src/common/bit.rs:137]: I'm curios why to read the raw pointer unaligned? its not guaranteed to be aligned?

File: native/core/src/common/bit.rs
Comment (comphead) [inline native/core/src/common/bit.rs:1038]: why is it u8 if we test u64, should we create a vector of i64?

Comment (kazuyukitanimura) [reply native/core/src/common/bit.rs:1038]: Because this is a buffer. It reads 8 bytes from here

File: native/core/src/common/bit.rs
Comment (kazuyukitanimura) [reply native/core/src/common/bit.rs:137]: hmm not sure if it is always aligned...

File: native/core/benches/bit_util.rs
Comment (kazuyukitanimura) [reply native/core/benches/bit_util.rs:189]: This is no a magic number, I just chose this randomly, no specific reasons

**Comment by kazuyukitanimura** [comment]
Plan to merge this today

**Comment by kazuyukitanimura** [comment]
Merged, thank you @andygrove @parthchandra @comphead

**Pull request merged**
